File: native-views/react-native-native-list/ios/NativeListLeadingVisual.swift
import UIKit

// Bounded visual primitive: three source slots, two optional overlays. It knows
// visual descriptors and request lifetimes, never row types or business keys.
final class NativeListLeadingVisual: UIView {
  var glyphSize: CGFloat = 18
  private let fallback = UILabel()
  private let icon = UIImageView()
  private let unread = UIView()
  private let networkBackdrop = UIView()
  private let corner = UIImageView()
  private let cornerBackdrop = UIView()
  private var slots: [NativeListImageSlot] = []
  private var overlaySlots: [NativeListImageSlot] = []
  private var overlays: [(UIView, [String: Any])] = []
  private let dashedBorder = CAShapeLayer()
  private var visual: [String: Any] = [:]
  private var imageStyle: [String: Any] = [:]
  private var sources: [([String: Any], String)] = []

  override init(frame: CGRect) {
    super.init(frame: frame)
    [fallback, icon, networkBackdrop, cornerBackdrop, corner, unread].forEach(addSubview)
    fallback.textAlignment = .center
    fallback.font = nativeListFont(ofSize: 13, weight: .bold)
    icon.contentMode = .scaleAspectFit
    corner.contentMode = .scaleAspectFit
    unread.backgroundColor = UIColor(nativeListHex: "#E5484D", fallback: .red)
    unread.layer.cornerRadius = 4
    networkBackdrop.layer.cornerRadius = 10
    cornerBackdrop.layer.cornerRadius = 10
    layer.addSublayer(dashedBorder)
  }
  required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

  func bind(
    _ visual: [String: Any], style: [String: Any], key: String, theme: [String: Any]?,
    isUnread: Bool
  ) {
    self.visual = visual
    imageStyle = style
    let kind = visual.string("kind")
    let variant =
      kind == "token"
      ? "token"
      : kind == "network" ? "network" : ["account", "wallet"].contains(kind) ? "avatar" : "generic"
    sources =
      kind == "stackedImages"
      ? visual.dictionaries("images").prefix(3).map { ($0, "generic") }
      : kind == "icon" ? [] : visual.dictionary("image").map { [($0, variant)] } ?? []
    if kind == "token", !sources.isEmpty, let network = visual.dictionary("networkImage") {
      sources.append((network, "network"))
    }
    while slots.count < sources.count {
      let slot = NativeListImageSlot()
      slots.append(slot)
      insertSubview(slot.view, belowSubview: slots.count == 1 ? networkBackdrop : cornerBackdrop)
    }
    let placeholder = theme?.string("strongBackground", default: "#0000000F") ?? "#0000000F"
    let fallbackIcon = visual.dictionary("fallbackIcon")
    let sourceFallback =
      kind != "icon" && !sources.isEmpty
      && sources[0].0.string("loadingStrategy", default: "none") != "none"
      && (visual["fallbackText"] != nil || fallbackIcon != nil)
    let fallbackKey = sourceFallback ? nativeListSourceFallbackStateKey(sources[0].0) : nil
    let restored = fallbackKey.map(NativeListSourceFallbackState.has) ?? false
    let fallbackData = kind == "icon" ? visual : fallbackIcon
    fallback.text = String(visual.string("fallbackText").prefix(2))
    fallback.textColor = UIColor(nativeListHex: "#8D8D8D", fallback: .gray)
    icon.image = fallbackData.map { nativeListIcon(named: $0.string("name")) } ?? nil
    icon.tintColor = UIColor(
      nativeListHex: fallbackData?.string("tintColor", default: "#646464") ?? "#646464",
      fallback: .darkGray)
    fallback.isHidden = kind == "icon" || fallbackIcon != nil || !sources.isEmpty && !restored
    icon.isHidden = !(kind == "icon" || (sources.isEmpty || restored) && fallbackIcon != nil)
    let visualBackground = UIColor(
      nativeListHex: visual.string(
        "backgroundColor", default: sources.isEmpty ? placeholder : "#00000000"), fallback: .clear)
    backgroundColor =
      sourceFallback ? UIColor(nativeListHex: placeholder, fallback: .clear) : visualBackground
    layer.borderWidth = kind == "icon" ? 1 / UIScreen.main.scale : 0
    layer.borderColor = UIColor(nativeListHex: "#0000001F", fallback: .clear).cgColor
    for (index, slot) in slots.enumerated() {
      guard index < sources.count else {
        slot.recycle()
        slot.view.isHidden = true
        continue
      }
      let source = sources[index]
      slot.view.isHidden = index > 0 || sourceFallback
      slot.bind(
        source.0, key: "\(key):leading:\(index)", variant: source.1,
        fit: index == 0 ? style["contentFit"] as? String : nil, placeholder: placeholder
      ) { [weak self, weak slot] loaded in
        slot?.view.isHidden = !loaded && (index > 0 || sourceFallback)
        guard let self, index == 0, sourceFallback else { return }
        if let fallbackKey {
          if loaded {
            NativeListSourceFallbackState.forget(fallbackKey)
          } else {
            NativeListSourceFallbackState.remember(fallbackKey)
          }
        }
        self.backgroundColor =
          loaded ? visualBackground : UIColor(nativeListHex: placeholder, fallback: .clear)
        self.fallback.isHidden = loaded || fallbackIcon != nil
        self.icon.isHidden = loaded || fallbackIcon == nil
      }
    }
    networkBackdrop.isHidden = !(kind == "token" && sources.count > 1)
    networkBackdrop.backgroundColor = nativeListColor(theme, "rowBackground", "#FFFFFF")
    let badge = visual.dictionary("cornerIcon")
    corner.isHidden = badge == nil
    cornerBackdrop.isHidden = badge == nil
    corner.image = badge.flatMap { nativeListIcon(named: $0.string("name")) }
    corner.tintColor = UIColor(
      nativeListHex: badge?.string("tintColor", default: "#646464") ?? "#646464",
      fallback: .darkGray)
    cornerBackdrop.backgroundColor = UIColor(
      nativeListHex: badge?.string("backgroundColor", default: "#FFFFFF") ?? "#FFFFFF",
      fallback: .white)
    unread.isHidden = !isUnread
    let previousOverlays = overlays
    overlays.removeAll(keepingCapacity: true)
    let descriptors = Array(visual.dictionaries("overlays").prefix(2))
    while overlaySlots.count < descriptors.count { overlaySlots.append(NativeListImageSlot()) }
    for (index, data) in descriptors.enumerated() {
      let frame = index < previousOverlays.count ? previousOverlays[index].0 : UIView()
      frame.clipsToBounds = true
      frame.backgroundColor = UIColor(
        nativeListHex: data.string(
          "backgroundColor",
          default: theme?.string("rowBackground", default: "#00000000") ?? "#00000000"),
        fallback: .clear)
      let child: UIView
      if let image = data.dictionary("image") {
        let slot = overlaySlots[index]
        child = slot.view
        child.isHidden = true
        slot.bind(image, key: "\(key):overlay:\(index)", placeholder: placeholder) {
          [weak child] loaded in child?.isHidden = !loaded
        }
      } else if !data.string("text").isEmpty {
        overlaySlots[index].recycle()
        let label = frame.subviews.first as? UILabel ?? UILabel()
        label.text = data.string("text")
        label.textAlignment = .center
        label.font = nativeListFont(ofSize: 10, weight: .medium)
        label.textColor = UIColor(
          nativeListHex: data.string("tintColor", default: "#646464"), fallback: .darkGray)
        child = label
      } else {
        overlaySlots[index].recycle()
        let glyph = frame.subviews.first as? UIImageView ?? UIImageView()
        glyph.image = nativeListIcon(named: data.string("name"))
        glyph.contentMode = .scaleAspectFit
        glyph.tintColor = UIColor(
          nativeListHex: data.string("tintColor", default: "#646464"), fallback: .darkGray)
        child = glyph
      }
      if child.superview !== frame {
        frame.subviews.forEach { $0.removeFromSuperview() }
        frame.addSubview(child)
      }
      if frame.superview == nil { addSubview(frame) }
      overlays.append((frame, data))
    }
    previousOverlays.dropFirst(descriptors.count).forEach { $0.0.removeFromSuperview() }
    for slot in overlaySlots.dropFirst(descriptors.count) { slot.recycle() }
    bringSubviewToFront(unread)
    setNeedsLayout()
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    let kind = visual.string("kind")
    let rtl = effectiveUserInterfaceLayoutDirection == .rightToLeft
    func logical(_ rect: CGRect) -> CGRect {
      rtl
        ? CGRect(x: bounds.width - rect.maxX, y: rect.minY, width: rect.width, height: rect.height)
        : rect
    }
    let shape = imageStyle.string(
      "shape", default: visual.string("shape", default: kind == "image" ? "rounded" : "circle"))
    let radius =
      imageStyle["cornerRadius"] != nil
      ? CGFloat(imageStyle.double("cornerRadius"))
      : shape == "square"
        ? 0
        : shape == "rounded"
          ? (imageStyle["shape"] != nil ? 10 : min(10, bounds.height / 4))
          : min(bounds.width, bounds.height) / 2
    layer.cornerRadius = radius
    clipsToBounds = sources.count < 2 && overlays.isEmpty && visual.dictionary("cornerIcon") == nil
    let ellipse = imageStyle.string("shape") == "circle" && imageStyle["cornerRadius"] == nil
    if ellipse && clipsToBounds {
      let mask = layer.mask as? CAShapeLayer ?? CAShapeLayer()
      mask.path = UIBezierPath(ovalIn: bounds).cgPath
      layer.mask = mask
    } else {
      layer.mask = nil
    }
    fallback.frame = bounds
    let fallbackIcon = visual.dictionary("fallbackIcon")
    let iconSize: CGFloat =
      !sources.isEmpty && fallbackIcon != nil
      ? min(bounds.width, bounds.height)
        * (fallbackIcon?.string("name") == "GlobusOutline" ? 1.2 : 1) : glyphSize
    icon.frame = CGRect(
      x: (bounds.width - iconSize) / 2, y: (bounds.height - iconSize) / 2, width: iconSize,
      height: iconSize)
    let tokenPair = kind == "token" && sources.count > 1
    for (index, slot) in slots.prefix(sources.count).enumerated() {
      let frame: CGRect
      if sources.count == 1 || tokenPair && index == 0 {
        frame = bounds
      } else if tokenPair {
        frame = CGRect(x: bounds.width - 14, y: bounds.height - 14, width: 16, height: 16)
      } else {
        let side = bounds.height * 0.72
        frame = CGRect(
          x: CGFloat(index) * (bounds.width - side) / CGFloat(max(1, sources.count - 1)),
          y: (bounds.height - side) / 2, width: side, height: side)
      }
      slot.view.frame = logical(frame)
      slot.view.clipsToBounds = true
      slot.view.layer.cornerRadius =
        index == 0 && (sources.count == 1 || tokenPair) ? radius : frame.height / 2
    }
    for (index, slot) in slots.enumerated() {
      if ellipse && index == 0 {
        let mask = slot.view.layer.mask as? CAShapeLayer ?? CAShapeLayer()
        mask.path = UIBezierPath(ovalIn: slot.view.bounds).cgPath
        slot.view.layer.mask = mask
      } else {
        slot.view.layer.mask = nil
      }
    }
    networkBackdrop.frame = logical(
      CGRect(x: bounds.width - 16, y: bounds.height - 16, width: 20, height: 20))
    cornerBackdrop.frame = networkBackdrop.frame
    corner.frame = cornerBackdrop.frame.insetBy(dx: 1, dy: 1)
    unread.frame = logical(CGRect(x: bounds.width - 8, y: 0, width: 8, height: 8))
    for (frame, data) in overlays {
      let size = CGFloat(data.double("size", default: 20))
      let width = CGFloat(data.double("width", default: Double(size)))
      let height = CGFloat(data.double("height", default: Double(size)))
      let offset = data.double("offset", default: 2)
      let x = CGFloat(data.double("offsetX", default: offset))
      let y = CGFloat(data.double("offsetY", default: offset))
      frame.frame = logical(
        CGRect(
          x: data.string("position") == "topLeft" ? -x : bounds.width - width + x,
          y: data.string("position") == "topLeft" ? -y : bounds.height - height + y, width: width,
          height: height))
      frame.layer.cornerRadius = min(width, height) / 2
      let inset = CGFloat(data.double("padding", default: 0))
      frame.subviews.first?.frame = frame.bounds.insetBy(
        dx: inset,
        dy: inset)
    }
    dashedBorder.isHidden = visual.string("borderStyle") != "dashed"
    dashedBorder.strokeColor =
      UIColor(nativeListHex: visual.string("borderColor", default: "#8D8D8D"), fallback: .gray)
      .cgColor
    dashedBorder.fillColor = UIColor.clear.cgColor
    dashedBorder.lineWidth = 2
    dashedBorder.lineDashPattern = [4, 4]
    dashedBorder.path = UIBezierPath(ovalIn: bounds.insetBy(dx: 1, dy: 1)).cgPath
  }
  func recycle() {
    slots.forEach { $0.recycle() }
    overlaySlots.forEach { $0.recycle() }
  }
}

File: native-views/react-native-native-list/android/src/main/java/com/onekey/nativelist/NativeListLeadingVisual.kt
package com.margelo.nitro.nativelist

import android.graphics.Color
import android.graphics.Outline
import android.graphics.drawable.GradientDrawable
import android.view.Gravity
import android.view.View
import android.view.ViewOutlineProvider
import android.widget.FrameLayout
import android.widget.TextView
import com.facebook.react.uimanager.ThemedReactContext
import kotlin.math.roundToInt
import org.json.JSONObject

/** Fixed visual slots; no row-template or business-key branching. */
internal class NativeListLeadingVisual(private val reactContext: ThemedReactContext) :
  FrameLayout(reactContext) {
  private val fallback =
    TextView(context).apply {
      gravity = Gravity.CENTER
      typeface = NativeListFonts.bold(context)
    }
  private val icon = OneKeyIconView(context)
  private val unread = View(context)
  private val networkBackdrop = View(context)
  private val corner = OneKeyIconView(context)
  private val cornerBackdrop = View(context)
  private val slots = mutableListOf<NativeListImageSlot>()
  private val overlaySlots = mutableListOf<NativeListImageSlot>()
  private val overlays = mutableListOf<Pair<FrameLayout, JSONObject>>()
  private var visual = JSONObject()
  private var style = JSONObject()
  private var sources = emptyList<Pair<JSONObject, String>>()
  val fallbackTextView: TextView
    get() = fallback

  var bitmapBorderWidth = 0

  /**
   * Legacy wallet-sidebar source-scale leading visual: text-only overlays are 16dp
   * tall pills with 12sp text on a 16dp line box, and a dashed frame border is 1dp.
   */
  var walletTextOverlays = false

  private fun isWalletText(data: JSONObject) =
    walletTextOverlays &&
      data.optString("text").isNotEmpty() &&
      data.optJSONObject("image") == null &&
      data.optString("name").isEmpty()

  private fun overlayHeight(data: JSONObject) =
    data.optInt("height", if (isWalletText(data)) 16 else data.optInt("size", 20))

  fun layoutNetworkBackdrop(top: Int, height: Int) {
    networkBackdrop.layout(networkBackdrop.left, top, networkBackdrop.right, top + height)
  }

  // Legacy: loaded images in a "rounded" slot clip at a fixed radius (10dp;
  // 8dp for explicit-height account selectors), independent of slot size.
  var roundedImageRadius = 10

  /** Market resolves its own radius; legacy still decoded circular bitmaps by shape. */
  var roundFollowsVisualShape = false
  var glyphSize = 18
  var roundedGlyphOrigin = false
  var iconBorder = true
  private var sourceScale = false

  private fun dp(value: Int) =
    if (sourceScale) (value * resources.displayMetrics.density).roundToInt()
    else NativeListScale.dp(resources, value)

  private fun color(value: String, fallback: Int = Color.TRANSPARENT) =
    runCatching { parseNativeListColor(value) }.getOrDefault(fallback)

  private fun fill(value: Int, radius: Float) =
    GradientDrawable().apply {
      setColor(value)
      cornerRadius = radius
    }

  init {
    listOf(fallback, icon, networkBackdrop, cornerBackdrop, corner, unread).forEach { addView(it) }
    clipChildren = false
    clipToPadding = false
  }

  fun bind(
    visual: JSONObject,
    style: JSONObject,
    key: String,
    theme: JSONObject?,
    isUnread: Boolean,
    sourceScale: Boolean,
    secondaryVisual: JSONObject? = null,
  ) {
    this.visual = visual
    this.style = style
    this.sourceScale = sourceScale
    icon.useSourceScale = sourceScale
    corner.useSourceScale = sourceScale
    val kind = visual.optString("kind")
    val variant =
      when (kind) {
        "token" -> "token"
        "network" -> "network"
        "account",
        "wallet" -> "avatar"
        else -> "generic"
      }
    sources =
      if (kind == "stackedImages")
        visual
          .optJSONArray("images")
          ?.let { images ->
            (0 until minOf(3, images.length())).mapNotNull {
              images.optJSONObject(it)?.let { image -> image to "generic" }
            }
          }
          .orEmpty()
      else if (kind == "icon") emptyList()
      else visual.optJSONObject("image")?.let { listOf(it to variant) }.orEmpty()
    if (kind == "token" && sources.isNotEmpty())
      visual.optJSONObject("networkImage")?.let { sources = sources + (it to "network") }
    secondaryVisual?.let { second ->
      val kind = second.optString("kind")
      val source =
        when (kind) {
          "stackedImages" -> second.optJSONArray("images")?.optJSONObject(0)
          "icon" -> null
          else -> second.optJSONObject("image")
        }
      if (source != null)
        sources =
          sources +
            (source to
              when (kind) {
                "token" -> "token"
                "network" -> "network"
                "account",
                "wallet" -> "avatar"
                else -> "generic"
              })
    }
    while (slots.size < sources.size) {
      val slot = NativeListImageSlot(reactContext)
      slots.add(slot)
      addView(slot.view, indexOfChild(if (slots.size == 1) networkBackdrop else cornerBackdrop))
    }
    val placeholder = theme?.optString("strongBackground", "#0000000F") ?: "#0000000F"
    val fallbackIcon = visual.optJSONObject("fallbackIcon")
    val sourceFallback =
      sources.firstOrNull()?.first?.optString("loadingStrategy", "none")?.let { it != "none" } ==
        true && (visual.has("fallbackText") || fallbackIcon != null)
    val fallbackKey =
      if (sourceFallback)
        sources.firstOrNull()?.first?.let { source ->
          nativeListSourceFallbackStateKey(
            source.optString("uri"),
            source
              .optJSONObject("headers")
              ?.let { headers ->
                headers.keys().asSequence().associateWith { headers.optString(it) }
              }
              .orEmpty(),
          )
        }
      else null
    val restored = fallbackKey?.let(NativeListSourceFallbackState::has) == true
    fallback.textSize = if (sourceScale) 13f else NativeListScale.font(resources, 13f)
    fallback.text = visual.optString("fallbackText").take(2)
    fallback.setTextColor(color("#00000072"))
    fallback.visibility =
      if (kind != "icon" && fallbackIcon == null && (sources.isEmpty() || restored)) VISIBLE
      else GONE
    val glyph = if (kind == "icon") visual else fallbackIcon
    icon.iconName = glyph?.optString("name") ?: ""
    icon.tintColor = color(glyph?.optString("tintColor", "#0000009B") ?: "#0000009B")
    icon.visibility =
      if (kind == "icon" || (sources.isEmpty() || restored) && fallbackIcon != null) VISIBLE
      else GONE
    val visualBackground =
      color(
        visual.optString("backgroundColor", if (sources.isEmpty()) placeholder else "#00000000")
      )
    val backgroundColor = if (sourceFallback) color(placeholder) else visualBackground
    background =
      fill(backgroundColor, 0f).apply {
        if (kind == "icon" && iconBorder) setStroke(1, color("#0000001F"))
        if (visual.optString("borderStyle") == "dashed")
          setStroke(
            dp(if (walletTextOverlays) 1 else 2),
            color(visual.optString("borderColor", "#00000072")),
            dp(4).toFloat(),
            dp(4).toFloat(),
          )
      }
    slots.forEachIndexed { index, slot ->
      if (index >= sources.size) {
        slot.recycle()
        slot.view.visibility = GONE
      } else {
        slot.view.visibility = if (index > 0 || sourceFallback) INVISIBLE else VISIBLE
        slot.bind(
          sources[index].first,
          "$key:leading:$index",
          sources[index].second,
          if (index == 0) style.optString("contentFit").takeIf { it.isNotEmpty() } else null,
          placeholder,
          round =
            visual.optString("shape", if (kind == "image") "rounded" else "circle") == "circle" &&
              (roundFollowsVisualShape || !style.has("shape") && !style.has("cornerRadius")),
        ) { loaded ->
          slot.view.visibility =
            if (!loaded && (index > 0 || sourceFallback)) INVISIBLE else VISIBLE
          if (index == 0 && sourceFallback) {
            fallbackKey?.let {
              if (loaded) NativeListSourceFallbackState.forget(it)
              else NativeListSourceFallbackState.remember(it)
            }
            (background as? GradientDrawable)?.setColor(
              if (loaded) visualBackground else color(placeholder)
            )
            fallback.visibility = if (!loaded && fallbackIcon == null) VISIBLE else GONE
            icon.visibility = if (!loaded && fallbackIcon != null) VISIBLE else GONE
          }
        }
      }
    }
    networkBackdrop.visibility = if (kind == "token" && sources.size > 1) VISIBLE else GONE
    networkBackdrop.background =
      fill(color(theme?.optString("rowBackground", "#FFFFFF") ?: "#FFFFFF"), dp(10).toFloat())
    val badge = visual.optJSONObject("cornerIcon")
    corner.visibility = if (badge == null) GONE else VISIBLE
    cornerBackdrop.visibility = corner.visibility
    corner.iconName = badge?.optString("name") ?: ""
    corner.tintColor = color(badge?.optString("tintColor", "#0000009B") ?: "#0000009B")
    cornerBackdrop.background =
      fill(
        color(
          badge?.optString(
            "backgroundColor",
            theme?.optString("rowBackground", "#FFFFFF") ?: "#FFFFFF",
          ) ?: "#FFFFFF"
        ),
        dp(10).toFloat(),
      )
    unread.visibility = if (isUnread) VISIBLE else GONE
    unread.background = fill(color("#E5484D"), dp(4).toFloat())
    val previousOverlays = overlays.toList()
    overlays.clear()
    val descriptors =
      visual
        .optJSONArray("overlays")
        ?.let { values -> (0 until minOf(2, values.length())).map { values.getJSONObject(it) } }
        .orEmpty()
    while (overlaySlots.size < descriptors.size) overlaySlots.add(NativeListImageSlot(reactContext))
    descriptors.forEachIndexed { index, data ->
      val frame = previousOverlays.getOrNull(index)?.first ?: FrameLayout(context)
      frame.background =
        fill(
          color(data.optString("backgroundColor", "#00000000")),
          minOf(
            dp(data.optInt("width", data.optInt("size", 20))),
            dp(overlayHeight(data)),
          ) / 2f,
        )
      frame.outlineProvider = ViewOutlineProvider.BACKGROUND
      frame.clipToOutline = true
      val child: View =
        data.optJSONObject("image")?.let { image ->
          val slot = overlaySlots[index]
          slot.view.visibility = INVISIBLE
          slot.bind(image, "$key:overlay:$index", placeholder = placeholder) {
            slot.view.visibility = if (it) VISIBLE else INVISIBLE
          }
          slot.view
        }
          ?: if (data.optString("text").isNotEmpty()) {
            overlaySlots[index].recycle()
            ((frame.getChildAt(0) as? TextView) ?: TextView(context)).apply {
              text = data.optString("text")
              gravity = Gravity.CENTER
              includeFontPadding = false
              val walletText = isWalletText(data)
              val size = if (walletText) 12f else 10f
              textSize = if (sourceScale) size else NativeListScale.font(resources, size)
              // The overlay TextView is reused: reset the line box for non-wallet text.
              if (walletText) androidx.core.widget.TextViewCompat.setLineHeight(this, dp(16))
              else setLineSpacing(0f, 1f)
              typeface =
                if (isWalletText(data)) NativeListFonts.regular(context)
                else NativeListFonts.medium(context)
              setTextColor(color(data.optString("tintColor", "#0000009B")))
            }
          } else {
            overlaySlots[index].recycle()
            ((frame.getChildAt(0) as? OneKeyIconView) ?: OneKeyIconView(context)).apply {
              useSourceScale = sourceScale
              iconName = data.optString("name")
              tintColor = color(data.optString("tintColor", "#0000009B"))
            }
          }
      val inset = data.optInt("padding", 0)
      if (isWalletText(data)) frame.setPadding(dp(2), 0, dp(2), 0)
      else frame.setPadding(dp(inset), dp(inset), dp(inset), dp(inset))
      if (child.parent !== frame) {
        frame.removeAllViews()
        // A reused slot view may still belong to a frame dropped by an earlier bind.
        (child.parent as? android.view.ViewGroup)?.removeView(child)
        frame.addView(child, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
      }
      if (frame.parent == null) addView(frame)
      overlays.add(frame to data)
    }
    previousOverlays.drop(descriptors.size).forEach { (frame, _) ->
      // Release the overlay child so its image slot can join a later frame.
      frame.removeAllViews()
      removeView(frame)
    }
    overlaySlots.drop(descriptors.size).forEach { it.recycle() }
    unread.bringToFront()
    requestLayout()
  }

  override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
    super.onLayout(changed, left, top, right, bottom)
    fun place(view: View, x: Int, y: Int, w: Int, h: Int) {
      val left = if (layoutDirection == LAYOUT_DIRECTION_RTL) width - x - w else x
      view.measure(
        MeasureSpec.makeMeasureSpec(w, MeasureSpec.EXACTLY),
        MeasureSpec.makeMeasureSpec(h, MeasureSpec.EXACTLY),
      )
      view.layout(left, y, left + w, y + h)
    }
    val w = width
    val h = height
    val kind = visual.optString("kind")
    val shape =
      style.optString(
        "shape",
        visual.optString("shape", if (kind == "image") "rounded" else "circle"),
      )
    val radius =
      if (style.has("cornerRadius"))
        (style.optDouble("cornerRadius") * resources.displayMetrics.density).toFloat()
      else
        when (shape) {
          "square" -> 0f
          "rounded" ->
            if (style.has("shape")) 10 * resources.displayMetrics.density
            else minOf(dp(10).toFloat(), h / 4f)
          else -> minOf(w, h) / 2f
        }
    (background as? GradientDrawable)?.apply {
      this.shape =
        if (style.optString("shape") == "circle" && !style.has("cornerRadius"))
          GradientDrawable.OVAL
        else GradientDrawable.RECTANGLE
      cornerRadius = radius
    }
    if (bitmapBorderWidth > 0) {
      (background as? GradientDrawable)?.setStroke(
        bitmapBorderWidth,
        color(visual.optString("borderColor")),
      )
    }
    place(fallback, 0, 0, w, h)
    val iconSize =
      if (sources.isNotEmpty() && visual.optJSONObject("fallbackIcon") != null)
        (minOf(w, h) *
            if (visual.optJSONObject("fallbackIcon")?.optString("name") == "GlobusOutline") 1.2f
            else 1f)
          .roundToInt()
      else dp(glyphSize)
    place(
      icon,
      (w - iconSize + if (roundedGlyphOrigin) 1 else 0) / 2,
      (h - iconSize + if (roundedGlyphOrigin) 1 else 0) / 2,
      iconSize,
      iconSize,
    )
    val tokenPair = kind == "token" && sources.size > 1
    slots.take(sources.size).forEachIndexed { index, slot ->
      val size =
        if (tokenPair && index == 1) dp(16)
        else if (sources.size == 1 || tokenPair) h else (h * 0.72f).roundToInt()
      val x =
        if (tokenPair && index == 1) w - size + dp(2)
        else if (sources.size == 1 || tokenPair) 0
        else (index * (w - size).toFloat() / maxOf(1, sources.size - 1)).roundToInt()
      val y = if (tokenPair && index == 1) h - size + dp(2) else (h - size) / 2
      val imageWidth = if (sources.size == 1 || tokenPair && index == 0) w else size
      slot.view.measure(
        MeasureSpec.makeMeasureSpec(imageWidth, MeasureSpec.EXACTLY),
        MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY),
      )
      val inset = if (index == 0) bitmapBorderWidth else 0
      place(slot.view, x + inset, y + inset, imageWidth - 2 * inset, size - 2 * inset)
      val imageRadius =
        if (shape == "rounded" && !style.has("shape") && !style.has("cornerRadius"))
          dp(roundedImageRadius).toFloat()
        else radius
      val r = if (index == 0 && (sources.size == 1 || tokenPair)) imageRadius else size / 2f
      slot.view.outlineProvider =
        object : ViewOutlineProvider() {
          override fun getOutline(view: View, outline: Outline) {
            if (index == 0 && style.optString("shape") == "circle" && !style.has("cornerRadius"))
              outline.setOval(0, 0, view.width, view.height)
            else outline.setRoundRect(-inset, -inset, view.width + inset, view.height + inset, r)
          }
        }
      slot.view.clipToOutline = true
    }
    place(networkBackdrop, w - dp(16), h - dp(16), dp(20), dp(20))
    place(cornerBackdrop, w - dp(16), h - dp(16), dp(20), dp(20))
    place(corner, w - dp(15), h - dp(15), dp(18), dp(18))
    place(unread, w - dp(8), 0, dp(8), dp(8))
    overlays.forEach { (frame, data) ->
      val size = data.optInt("size", 20)
      val oh = dp(overlayHeight(data))
      val ow =
        if (isWalletText(data) && !data.has("width")) {
          // Legacy: wallet text pills wrap their label plus horizontal padding.
          val child = frame.getChildAt(0)
          child?.measure(
            MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED),
            MeasureSpec.makeMeasureSpec(oh, MeasureSpec.EXACTLY),
          )
          (child?.measuredWidth ?: 0) + frame.paddingLeft + frame.paddingRight
        } else dp(data.optInt("width", size))
      val x = dp(data.optInt("offsetX", data.optInt("offset", 2)))
      val y = dp(data.optInt("offsetY", data.optInt("offset", 2)))
      val ox = if (data.optString("position") == "topLeft") -x else w - ow + x
      val oy = if (data.optString("position") == "topLeft") -y else h - oh + y
      frame.measure(
        MeasureSpec.makeMeasureSpec(ow, MeasureSpec.EXACTLY),
        MeasureSpec.makeMeasureSpec(oh, MeasureSpec.EXACTLY),
      )
      place(frame, ox, oy, ow, oh)
    }
  }

  fun recycle() {
    slots.forEach { it.recycle() }
    overlaySlots.forEach { it.recycle() }
  }

  fun dispose() {
    slots.forEach { it.dispose() }
    overlaySlots.forEach { it.dispose() }
  }
}
